File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.js
function solution(operations) {
    const answer = [];
    operations.map(v => {
        v = v.split(" ");
        
        if (v[0]==='I') answer.push(parseInt(v[1]));
        if (v[0]==='D') {
            if (v[1]==='1') answer.splice(answer.indexOf(Math.max(...answer)), 1);
            else answer.splice(answer.indexOf(Math.min(...answer)),1);
        }
    });
    
    return answer.length===0 ? [0,0] : [Math.max(...answer),Math.min(...answer)];
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq
def solution(operations):
	answer = []
	
	min_heap = []
	max_heap = []
	num = {} # 숫자 개수 저장
	
	for oper in operations:
		t,n= oper.split()
		n = int(n)
		
		if t=="I":
			heapq.heappush(min_heap, n)
			heapq.heappush(max_heap, -n)
			if num.get(n):
				num[n]+=1
			else:
				num[n] = 1
				
		else:
			if n==1:
				if len(max_heap):
					add = -heapq.heappop(max_heap)
				else:
					continue
				while max_heap and num[add]<1:
					add = -heapq.heappop(max_heap)
				num[add] -= 1
				
			else:
				if len(min_heap):
					add = heapq.heappop(min_heap)
				else:
					continue
				while min_heap and num[add]<1:
					add = heapq.heappop(min_heap)
				num[add] -= 1
				
	while min_heap and num[min_heap[0]]<1:
		heapq.heappop(min_heap)
	min = min_heap[0] if len(min_heap) else 0
		
	while max_heap and num[-max_heap[0]]<1:
		heapq.heappop(max_heap)
	max = -max_heap[0] if len(max_heap) else 0
		
	return [max, min]